fpm/src/fpm.f90

```diff
@@ -126,93 +144,337 @@ subroutine cmd_build(settings)
+      call warnwrite(join_path(settings%name, 'app/main.f90'), littlefile)
    endif
 
-   write(*,'(a)')message
-   print *, "fpm error: 'fpm new' not implemented."
-   error stop 1
-end subroutine cmd_new
+   call warnwrite(join_path(settings%name, 'fpm.toml'), message)               ! now that built it write NAME/fpm.toml
 
-subroutine cmd_run(settings)
-    type(fpm_run_settings), intent(in) :: settings
-    integer                      :: i
-
-    write(*,*)'RELEASE=',settings%release
-    if(size(settings%name).eq.0)then
-        write(*,*)'RUN DEFAULTS with arguments ['//settings%args//']'
-    else
-        do i=1,size(settings%name)
-            write(*,*)'RUN:'//trim(settings%name(i))//' with arguments ['//settings%args//']'
-        enddo
-    endif
-
-    print *, "fpm error: 'fpm run' not implemented."
-    error stop 1
+   call run('cd ' // settings%name // ';git init')    ! assumes these commands work on all systems and git(1) is installed 
+contains
+!===================================================================================================================================
+subroutine warnwrite(fname,data)
+character(len=*),intent(in) :: fname
+character(len=*),intent(in) :: data(:)
+   if(.not.exists(fname))then
+      call filewrite(fname,data)
+   else
+      write(stderr,'(*(g0,1x))')'fpm::new<WARNING>',fname,'already exists. Not overwriting'
+   endif
+end subroutine warnwrite
+!===================================================================================================================================
+subroutine filewrite(filename,filedata)
+use,intrinsic :: iso_fortran_env, only : stdin=>input_unit, stdout=>output_unit, stderr=>error_unit
+! write filedata to file filename
+character(len=*),intent(in)           :: filename
+character(len=*),intent(in)           :: filedata(:)
+integer                               :: lun, i, ios
+character(len=256)                    :: message
+   message=' '
+   ios=0
+   if(filename.ne.' ')then
+    open(file=filename, &
+    & newunit=lun, &
+    & form='formatted', &      !  FORM      =  FORMATTED   |  UNFORMATTED
+    & access='sequential', &   !  ACCESS    =  SEQUENTIAL  |  DIRECT       |  STREAM
+    & action='write', &        !  ACTION    =  READ|WRITE  |  READWRITE
+    & position='rewind', &     !  POSITION  =  ASIS        |  REWIND       |  APPEND
+    & status='new', &          !  STATUS    =  NEW         |  REPLACE      |  OLD     |  SCRATCH   | UNKNOWN
+    & iostat=ios, &
+    & iomsg=message)
+   else
+      lun=stdout
+      ios=0
+   endif
+   if(ios.ne.0)then
+      write(stderr,'(*(a,1x))')'*filewrite* error:',filename,trim(message)
+      error stop 1
+   endif
+   do i=1,size(filedata)                                                    ! write file
+      write(lun,'(a)',iostat=ios,iomsg=message)trim(filedata(i))
+      if(ios.ne.0)then
+         write(stderr,'(*(a,1x))')'*filewrite* error:',filename,trim(message)
+         stop 4
```

Comment:
Any reason to use stop here and error stop in the other places?

Reply:
The original skeleton had error stop, and there is some advantage to using an exit code to allow for unit testing. For gfortran an error stop also produces a traceback which can be useful for development but since gfortran and some other compilers produce a distracting "STOP" message it is  very ugly for an interactive utility. In a production release I would recommend a standard stop everywhere except where an unexpected stop is provided. Unfortunately Fortran does not provide an equivalent to exit() that just sets the status. If you use the program interactively it only takes one stop error or a few stop N uses to find the resulting output is very distracting. A traceback can be a great tool for development but seems a bit hysterical for typical events that have been described with an error message already. So at this point I would recommend a custom stop function be called with a number that can later be turned off or only activated in test mode unless an exit(3c) is added to the stdlib. So a stop with no value was intentional. Would you consider a non-zero exit critical?

Reply:
Basically the error stop was already being used in the original skeleton I started with and I have no strong preference on how stop is used except I hate there being no way to surpress the "STOP" message in gfortran (that varies from compiler to compiler). If anyone feels we need a rule like "error stop" for unexpected  internal errors; all stop should have a numeric value, etc. let me know and I think it should then be in the style guide. Why on earth does the standard not allow for a stop with a string and a value where the string can be null? The differences between compilers regarding what output is produced by STOP is irrirating. And an explicit traceback or abort routine is preferable to leaving that up to each implementor to.